src/Toolchain.ml

```diff
@@ -0,0 +1,647 @@
+             ; lsp = (fun () -> (name, [| "exec"; "ocamllsp" |]))
+             }
+             |> R.return)
+  end
+
+  let make ~env ~root ~discoveredManifestPath ~t =
+    match t with
+    | Opam -> Opam.make ~env ~root ~discoveredManifestPath
+    | Esy -> Esy.make ~env ~root ~discoveredManifestPath
+
+  let toName = function
+    | Opam -> "opam"
+    | Esy -> "esy"
+
+  let ofName = function
+    | "opam" -> Ok Opam
+    | "esy" -> Ok Esy
+    | n -> Error {j|Invalid name $n was provided|j}
+
+  let specOfName ~env ~name ~root ~discoveredManifestPath =
+    match name with
+    | x when x = Opam.name -> Opam.make ~env ~root ~discoveredManifestPath
+    | x when x = Esy.name -> Esy.make ~env ~root ~discoveredManifestPath
+    | _ -> "Invalid package manager name" |> R.fail |> Js.Promise.resolve
+
+  module PackageManagerSpecTuple = struct
+    type t = a * Fpath.t
+
+    let compare x y =
+      let xa, xb = x in
+      let ya, yb = y in
+      if xa = ya && xb = yb then
+        0
+      else
+        1
+  end
+
+  module PackageManagerSpecTupleSet = Set.Make (PackageManagerSpecTuple)
+
+  let alreadyUsed folder =
+    [ Esy; Opam ] |> Array.of_list
+    |> Array.map (fun pm ->
+           let lockFileFpath =
+             match pm with
+             | Opam ->
+               let open Fpath in
+               join folder Opam.lockFile
+             | Esy ->
+               let open Fpath in
+               join folder Esy.lockFile
+           in
+           Fs.exists (Fpath.show lockFileFpath)
+           |> Js.Promise.then_ (fun exists -> Js.Promise.resolve (pm, exists)))
+    |> Js.Promise.all
+    |> Js.Promise.then_ (fun l ->
+           l
+           |> Js.Array.filter (fun (_pm, used) -> used)
+           |> Array.map (fun t ->
+                  let pm, _used = t in
+                  pm)
+           |> Array.to_list |> R.return |> Js.Promise.resolve)
+
+  let available ~env =
+    supportedPackageManagers
+    |> List.map (fun (pm : t) ->
+           let name =
+             match pm with
+             | Opam -> Opam.name
+             | Esy -> Esy.name
+           in
+           Cmd.make ~env ~cmd:name
+           |> Js.Promise.then_ (fun r ->
+                  let r =
+                    match r with
+                    | Ok _ -> (pm, true)
+                    | Error _e -> (pm, false)
+                  in
+                  Js.Promise.resolve r))
+    |> Array.of_list |> Js.Promise.all
+    |> Js.Promise.then_ (fun r ->
+           Js.Array.filter (fun (_pm, available) -> available) r
+           |> Array.map (fun (pm, _used) -> pm)
+           |> Array.to_list |> R.return |> Js.Promise.resolve)
+
+  let setupToolChain spec = spec.setup ()
+
+  let lsp spec = spec.lsp ()
+
+  let env spec = spec.env ()
+
+  let globalSpec env =
+    { setup = (fun () -> Js.Promise.resolve (Ok ()))
+    ; lsp = (fun () -> ("ocamllsp", [||]))
+    ; env = (fun () -> env |> R.return |> Js.Promise.resolve)
```

Comment:
The use of callbacks here is completely unjustified. As far as I can tell, there are basically 3 possible values for all these fields. Why do we need to pass functions then?